### PR TITLE
Expanding Scripting API: AddBuff

### PR DESCRIPTION
**Pull request by abosma**
Combined the code for AddBuff and AddBuffHudVisual so it's not needed to have them be separated. Haven't removed AddBuffHudVisual atm since lot's of scripts rely on it still.

### CODE REVIEW

File: GameServerLib/API/ApiFunctionManager.cs [hunk 1]
```diff
@@ -116,11 +116,26 @@ public static bool IsWalkable(float x, float y)
             return _game.Map.NavGrid.IsWalkable(x, y);
         }
 
-        public static void AddBuff(string buffName, float duration, byte stacks, BuffType buffType, IObjAiBase onto, IObjAiBase from)
+        public static void AddBuff(string buffName, float duration, byte stacks, BuffType buffType, IObjAiBase onto, IObjAiBase from, float removeAfter = -1.0f)
         {
+            if (duration < 0)
```
Comment (lzardy) [inline GameServerLib/API/ApiFunctionManager.cs]: This should be in Buff.cs, line 57.

File: GameServerLib/API/ApiFunctionManager.cs [hunk 2]
```diff
@@ -116,11 +116,26 @@ public static bool IsWalkable(float x, float y)
...
+            {
+                return;
+            }
+
             var buff = new Buff(_game, buffName, duration, stacks, buffType, onto, from);
+
             onto.AddBuff(buff);
```
Comment (lzardy) [inline GameServerLib/API/ApiFunctionManager.cs:133]: If you are fine with changing the title of this PR to something like "Expanding Scripting API: AddBuff", you could add to this function the ability to call AddBuffGameScript.
You will need to do:
`public static void AddBuff(string buffName, float duration, byte stacks, BuffType buffType, ISpell spell, IObjAiBase onto, IObjAiBase from, float removeAfter = -1.0f, bool isUnique = false)`
then after `onto.AddBuff(buff);` you will need to add `onto.AddBuffGameScript(buffName, buffName, spell, removeAfter, isUnique);`

Comment (lzardy) [reply GameServerLib/API/ApiFunctionManager.cs:133]: Could also probably remove all of `removeAfter` and replace with `duration`.

File: GameServerLib/API/ApiFunctionManager.cs [hunk 3]
```diff
@@ -118,9 +118,41 @@ public static bool IsWalkable(float x, float y)
...
+
             _game.PacketNotifier.NotifyAddBuff(buff);
+
+            CreateTimer(duration, () =>
```
Comment (danil179) [inline GameServerLib/API/ApiFunctionManager.cs:138]: Add here `// HACK: Remove timers and use proper tick system to handle the buffs using priority queue of duration and order`

Comment (abosma) [reply GameServerLib/API/ApiFunctionManager.cs:138]: Added.

File: GameServerLib/API/ApiFunctionManager.cs [hunk 4]
```diff
@@ -118,9 +118,43 @@ public static bool IsWalkable(float x, float y)
 
         public static void AddBuff(string buffName, float duration, byte stacks, BuffType buffType, IObjAiBase onto, IObjAiBase from)
         {
+            if (duration <= 0)
```
Comment (danil179) [inline GameServerLib/API/ApiFunctionManager.cs]: Move the validity checks into `Buff` class.

**Review by danil179 — CHANGES_REQUESTED** [review]
Some checks should be more internal.

**Comment by abosma** [comment]
Buff now returns an ArgumentException if duration is 0 or less.

**Comment by lzardy** [comment]
> Buff now returns an ArgumentException if duration is 0 or less.

The point of returning rather than having an argumentexception is that it will simply make the buff last forever, this means we don't need to use arbitrarily large numbers for infinitely long buffs.

**Review by danil179 — APPROVED** [review]
Approved.

For infinite long buffs one should use special flag.

**Review by danil179 — APPROVED** [review]
It will possibly break so many scripts (because of the 'usual' negative duration), so we need to label this PR as "potentially-script-breaking".

**Comment by lzardy** [comment]
None of the current scripts use AddBuff or AddBuffGameScript from the ApiFunctionManager, they all use AddBuffGameScript from ObjAiBase. They will work fine.

**Pull request merged**